openedx_events/tooling.py

```diff
@@ -0,0 +1,199 @@
+"""
+Tooling necessary to use Open edX events.
+"""
+import warnings
+from datetime import datetime
+
+from django.conf import settings
+from django.dispatch import Signal
+
+import openedx_events
+from openedx_events.exceptions import InstantiationError, SenderValidationError
+
+try:
+    import crum
+except ImportError:
+    crum = None
+
+
+class OpenEdxPublicSignal(Signal):
+    """
+    Custom class used to create Open edX events.
+    """
+
+    def __init__(self, event_type, data, minor_version="0.0"):
+        """
+        Init method for OpenEdxPublicSignal definition class.
+
+        Arguments:
+            event_type (str): name of the event.
+            data (dict): attributes passed to the event.
+            minor_version (str): version of the event type.
+        """
+        if not event_type:
+            raise InstantiationError(
+                message="Missing required argument 'event_type'"
+            )
+        if not data:
+            raise InstantiationError(
+                event_type=event_type, message="Missing required argument 'data'"
+            )
+        self.init_data = data
+        self.event_type = event_type
+        self.minor_version = minor_version
+        super().__init__()
+
+    def __repr__(self):
+        """
+        Represent OpenEdxPublicSignal as a string.
+        """
+        return "<OpenEdxPublicSignal: {event_type}>".format(event_type=self.event_type)
+
+    def get_signal_metadata(self):
```

Comment:
A few comments:

**Defining Metadata in data.py**
Can the metadata itself be another `attrs` class that has well defined fields? This could live somewhere independently of the learning events.

**Separating CloudEvents serialization specifics.**
Let's keep the metadata here in a more natural Python form while it's in memory, and convert it to the serialized version in an explicit CloudEvents layer in the future, when that becomes relevant. In concrete terms, I think that just means that `specversion` goes away at this layer (it's a fixed constant for CloudEvents), time becomes a UTC `datetime` instead of a string, and events version might become a tuple of ints so that version comparisons don't have to worry about lexical ordering of strings (e.g. '0.9.0' vs. '0.10.0').

**Message Queue version**
_This isn't something I think you need to do yet_, but is the long term plan to make `metadata` an optional param that's auto-generated in the case of a local send, but populated from a serialized message if the signal is transmitted over a message queue?

**UUID?**
The CloudEvent will eventually have a UUID. Is it useful to have it in this metadata as well?

**Nit: "get" vs. "create/generate"**
This is a nit, but for me, `get` implies something is more or less staying the same and you're accessing it, while `create` or `generate` means that you're creating a new thing and handing it back.

Reply:
Hey there @ormsbee! I hope you're doing good!

I've updated the PR with your recommendations:

**Defining Metadata in data.py**
I've created [this](https://github.com/eduNEXT/openedx-events/pull/17/files) PR showing how I would do this

**Separating CloudEvents serialization specifics.**
I followed all your recommendations here.

**Message Queue version**
I think the best thing is to consider this use case already. If a Metadata object exists, then it's not generated. What do you think? 

**UUID?**
I added an ID as part of the metadata object, based on what the OEP-41 said:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0041-arch-async-server-event-messaging.html#id2


**Nit: "get" vs. "create/generate"**
Thanks for the comment! I fixed it.

Reply:
Update: we fully adopted defining metadata in data.py